py/obj: Explicitly cast MP_DEFINE_CONST_OBJ_TYPE slot values.

AJMansfield opened the pull request:
### Summary
This PR updates the definition of `MP_DEFINE_CONST_OBJ_TYPE` to generate a compile-time error on code that attempts to define a slot function with an incompatible function signature, and rewrites the code generation script that produces its internal `MP_DEFINE_CONST_OBJ_TYPE_NARGS_*` macros to make it easier to read and modify in this way.

The addition of the explicit `(const void *)` casts also resolves an issue compiling with IAR, where an implicit cast from a function pointer to a void pointer is disallowed, based on comments from @hammondeggs in the MicroPython discord from [this conversation](https://discord.com/channels/574275045187125269/1012726673709469818/1423683501374308392).

### Testing
I've built and run the test suite against the Unix and RP2 ports with these patches applied, with all tests passing.

To verify that this generates the expected error, I've also tested building a defective version of MicroPython where I've added an `attr` slot method to the objbool type with an invalid signature: `static void bool_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest, mp_obj_t extra)`.

Applied to master, the aforementioned defect produces no compilation warnings or test failures! But with this PR, the build correctly produces the following error identifying the problem:

```
In file included from ../../py/mpstate.h:35,
                 from ../../py/runtime.h:29,
                 from ../../py/objbool.c:29:
../../py/obj.h:771:37: error: cast between incompatible function types from ‘void (*)(void *, qstr,  void **, void *)’ {aka ‘void (*)(void *, long unsigned int,  void **, void *)’} to ‘void (*)(void *, qstr,  void **)’ {aka ‘void (*)(void *, long unsigned int,  void **)’} [-Werror=cast-function-type]
  771 | #define _MP_OBJ_TYPE_SLOT_TYPE_attr (mp_attr_fun_t)
      |                                     ^
../../py/obj.h:821:44: note: in definition of macro ‘MP_DEFINE_CONST_OBJ_TYPE_EXPAND’
  821 | #define MP_DEFINE_CONST_OBJ_TYPE_EXPAND(x) x
      |                                            ^
../../py/obj.h:811:533: note: in expansion of macro ‘_MP_OBJ_TYPE_SLOT_TYPE_attr’
  811 | #define MP_DEFINE_CONST_OBJ_TYPE_NARGS_5(_struct_type, _typename, _name, _flags, f1, v1, f2, v2, f3, v3, f4, v4, f5, v5) const _struct_type _typename = { .base = { &mp_type_type }, .flags = _flags, .name = _name, .slot_index_##f1 = 1, .slot_index_##f2 = 2, .slot_index_##f3 = 3, .slot_index_##f4 = 4, .slot_index_##f5 = 5, .slots = { (const void *)_MP_OBJ_TYPE_SLOT_TYPE_##f1 v1, (const void *)_MP_OBJ_TYPE_SLOT_TYPE_##f2 v2, (const void *)_MP_OBJ_TYPE_SLOT_TYPE_##f3 v3, (const void *)_MP_OBJ_TYPE_SLOT_TYPE_##f4 v4, (const void *)_MP_OBJ_TYPE_SLOT_TYPE_##f5 v5 } }
      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:825:179: note: in expansion of macro ‘MP_DEFINE_CONST_OBJ_TYPE_NARGS_5’
  825 | #define MP_DEFINE_CONST_OBJ_TYPE_NARGS(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, N, ...) MP_DEFINE_CONST_OBJ_TYPE_NARGS_##N
      |                                                                                                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:834:71: note: in expansion of macro ‘MP_DEFINE_CONST_OBJ_TYPE_NARGS’
  834 | #define MP_DEFINE_CONST_OBJ_TYPE(...) MP_DEFINE_CONST_OBJ_TYPE_EXPAND(MP_DEFINE_CONST_OBJ_TYPE_NARGS(__VA_ARGS__, _INV, 12, _INV, 11, _INV, 10, _INV, 9, _INV, 8, _INV, 7, _INV, 6, _INV, 5, _INV, 4, _INV, 3, _INV, 2, _INV, 1, _INV, 0)(mp_obj_type_t, __VA_ARGS__))
      |                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/objbool.c:91:1: note: in expansion of macro ‘MP_DEFINE_CONST_OBJ_TYPE’
   91 | MP_DEFINE_CONST_OBJ_TYPE(
      | ^~~~~~~~~~~~~~~~~~~~~~~~
CC ../../py/objgenerator.c
cc1: all warnings being treated as errors
```

(For an error generated this deep within a nested macro expansion, it's shockingly readable --- it even indicates that it's the `attr` slot that's the issue!)

I lack the infrastructure to test IAR builds myself, but @hammondeggs has reported that this PR does resolve their issue.

### Trade-offs and Alternatives
There are no downsides to this change --- this refactor doesn't affect the compiled output, it just improves compiler diagnostics and compatibility. With the refactor to the script generating them, it's also an improvement to code clarity even despite making the nargs macros themselves quite a bit wordier.

AJMansfield committed:
py/obj: Reorganize slot access macros to put with slot definitions.

i.e. rather than having them just arbitrarily between the generated
macros for the define macro and the other macros in that suite.

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

AJMansfield committed:
py/obj: Explicitly cast slot initializer values to `(const void *)`.

This resolves an issue compiling with IAR, where an implicit cast from a
function pointer to a void pointer is disallowed.

Based on comments from @hammondeggs in the MicroPython discord:
https://discord.com/channels/574275045187125269/1012726673709469818/1423683501374308392

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

AJMansfield committed:
py/obj: Enforce slot function signatures at compile time.

By explicitly casting each slot to its slot type, this will ensure that
a compiler diagnostic is generated when MP_DEFINE_CONST_OBJ_TYPE is
invoked with a slot function with the incorrect signature.

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

AJMansfield commented:
unix/qemu_arm test failure on 568fbb43268368a12e6ac5c1e08dd68867a41746 is spurious (`thread/stress_heap.py`).

hammondeggs commented:
This has fixed this issue that I ran into when attempting to compile in IAR. Thank you!

AJMansfield committed:
py/obj: Eliminate trailing comma from MP_DEFINE_CONST_OBJ_TYPE_NARGS.

This is mostly to simplify and make the associated code-gen logic easier
to modify, by inserting the comma in the "obvious" place as part of the
`.join` rather than as part of the element template.

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

AJMansfield committed:
py/obj: Refactor MP_DEFINE_CONST_OBJ_TYPE_NARGS script for readability.

This is a comment-only change. The edited script produces exactly the
same output as the original, just in a way that will allow more minimal
diffs to modify it.

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

projectgus reviewed (approved):
Catching this mismatch at compile time is very useful and the changes look good to me,, thanks @AJMansfield!

dpgeorge reviewed (approved):
Thanks for splitting this change into commits that are easy to review.
